### ISSUE TITLE
 Better testing of SQL to improve it's stability

### ISSUE THREAD

**Issue by ligurio** [tarantool/tarantool#4826]
SQL is a part of Tarantool and stability of it closely connected to stability of SQL.
We have a numerous bugs related to SQL subsystem stability (parser, lexer, ).
We need to decrease probability of such bugs and mitigate them in a tarantool release cycle. I propose following tools and plan:

1. `sqlsmith` or any other SQL grammar fuzzer
Tarantool uses SQL engine imported from SQLite project. Taking into account perfect testing in SQLite (see https://www.sqlite.org/testing.html) I believe fuzzing SQL engine itself will be useless. We should focus on testing integration layer of SQL engine with tarantool core and error handling.
2. LibFuzzer: analyze sql bugs and find functions in source code where these bugs were fixed. Create focused fuzzers for these functions using LibFuzzer.
3. emulate OOM? (see #3150)

### References

1. https://github.com/tarantool/tarantool/issues?q=is%3Aissue+is%3Aopen+label%3Acrash+label%3Asql+

**Comment by rybakit** [comment]
A bit off-topic: since you mentioned fuzzing, would it also make sense to test the msgpack library used in Tarantool with the fuzzer? At least it helped me to find some bugs in two msgpack libraries (not related to Tarantool).

**Comment by ligurio** [comment]
> A bit off-topic: since you mentioned fuzzing, would it also make sense to test the msgpack library used in Tarantool with the fuzzer? At least it helped me to find some bugs in two msgpack libraries (not related to Tarantool).

Sure, but it is out of scope of current ticket.

**Comment by tsafin** [comment]
Agreed that having fuzzer as part of testing of any subsystem is a must. But apparently there is bigger missing parts we could get benefit of:
- like sqllogictest which is not ported here, as far as I know;
- or speedtest1

I believe that proper set of sqllogictest  tests could provide us better coverage than random fuzzer. Which will give us better stability in a longer run.

**Comment by ligurio** [comment]
@tsafin Timur, thanks for you comment! Regarding speedtest1 and sqllogictest tests:
I read description for both of them and looks like speedtest1 is a sqlite-specific benchmark (i.e. *performance* measurement) and sqllogictest is a tool for testing *correctness* of SQL queries. Both tests are interesting in scope of tarantool testing but it is out of scope current task.

**Comment by ligurio** [comment]
Update regarding SQL fuzzing:

- cannot build sqlsmith because it always requires libpxx library even postgresql support disabled, looks like a bug in sqlsmith
- prepared [a draft version of SQL grammar]( - https://github.com/ligurio/grammars/blob/master/sqlite.lark) (sqlite's dialect). Generation of a single SQL query tooks a lot of time - 1min.
- tried [sqltest](https://github.com/elliotchance/sqltest), it is written on Python2 and requires driver for Tarantool, but looks like tarantool-python cannot execute SQL statements and tnt-async written on Python 3.

**Comment by ligurio** [comment]
[APOLLO: Automatic Detection and Diagnosis ofPerformance Regressions in Database Systems](http://www.vldb.org/pvldb/vol13/p57-jung.pdf)
